src/index.ts

```diff
@@ -0,0 +1,326 @@
+   * List of recently used block types
+   */
+  private recentlyUsedBlocks: string[] = [];
+  /**
+   * Checks if the workspace is finished loading, to avoid taking action on
+   * all the BLOCK_CREATE events during workspace loading.
+   */
+  /* eslint-disable-next-line @typescript-eslint/explicit-member-accessibility */
+  public workspaceHasFinishedLoading = false;
+
+  /**
+   * Config parameter which sets the size of the toolbox categories.
+   */
+  private numBlocksPerCategory: number;
+
+  /**
+   * Constructs a BlockSuggestor object.
+   *
+   * @param numBlocksPerCategory the size of each toolbox category
+   */
+  constructor(numBlocksPerCategory: number) {
+    /**
+     * Config parameter which sets the size of the toolbox categories.
+     */
+    this.numBlocksPerCategory = numBlocksPerCategory;
+
+    // Bind methods to `this`
+    this.eventListener = this.eventListener.bind(this);
+    this.getMostUsed = this.getMostUsed.bind(this);
+    this.getRecentlyUsed = this.getRecentlyUsed.bind(this);
+    this.generateBlockData = this.generateBlockData.bind(this);
+  }
+
+  /**
+   * Sets the machine learning model to be used for block suggestions.
+   *
+   * @param model - The machine learning model to set
+   */
+  setModel(model: IMachineLearningModel): void {
+    this.model = model;
+  }
+
+  /**
+   * Generates block suggestions based on the current description.
+   */
+  getSuggestedBlocks = (): Blockly.utils.toolbox.BlockInfo[] => {
+    const description =
+      typeof this.description === 'string' ? this.description : '';
+
+    const suggestedBlocks = this.model
+      ? this.model.getSuggestedBlocks(description)
+      : [];
+
+    const blockTypes = suggestedBlocks.map((block) => block.type);
+    return this.generateBlockData(blockTypes);
+  };
+
+  /**
+   * Generates a list of the 10 most frequently used blocks, in order.
+   * Includes a secondary sort by most recent blocks.
+   *
+   * @returns A list of block JSON
+   */
+  getMostUsed = (): Blockly.utils.toolbox.BlockInfo[] => {
+    // Store the frequency of each block, as well as the index first appears at.
+    const countMap = new Map<string, number>();
+    const recencyMap = new Map<string, number>();
+
+    this.recentlyUsedBlocks.forEach((key, index) => {
+      countMap.set(key, (countMap.get(key) || 0) + 1);
+      if (!recencyMap.has(key)) {
+        recencyMap.set(key, index + 1);
+      }
+    });
+
+    // Get a sorted list.
+    const freqUsedBlockTypes = Array.from(countMap.keys()).sort(
+      (a, b) =>
+        countMap.get(b) -
+        countMap.get(a) +
+        0.01 * (recencyMap.get(a) - recencyMap.get(b)),
+    );
+
+    return this.generateBlockData(freqUsedBlockTypes);
+  };
+
+  /**
+   * Generates a list of the 10 most recently used blocks.
+   *
+   * @returns A list of block JSON objects
+   */
+  getRecentlyUsed = (): Blockly.utils.toolbox.BlockInfo[] => {
+    const uniqueRecentBlocks = Array.from(new Set(this.recentlyUsedBlocks));
```

Comment:
Change this back to the more concise form it originally was:
```suggestion
    const uniqueRecentBlocks = [...new Set(this.recentlyUsedBlocks)];
```